File: DSA/Arshdeep-Kaur/Day15_Question2.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <sstream>
#include <string>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

vector<int> levelOrder(TreeNode *root)
{
    vector<int> result;
    if (!root)
        return result;

    queue<TreeNode *> q;
    q.push(root);

    while (!q.empty())
    {
        TreeNode *node = q.front();
        q.pop();
        result.push_back(node->val);

        if (node->left)
            q.push(node->left);
        if (node->right)
            q.push(node->right);
    }

    return result;
}

TreeNode *buildTree(const string &data)
{
    if (data.empty())
        return nullptr;

    istringstream iss(data);
    string val;
    iss >> val;

    TreeNode *root = new TreeNode(stoi(val));
    queue<TreeNode *> q;
    q.push(root);

    while (iss >> val)
    {
        TreeNode *parent = q.front();
        q.pop();

        if (val != "null")
        {
            parent->left = new TreeNode(stoi(val));
            q.push(parent->left);
        }

        if (iss >> val && val != "null")
        {
            parent->right = new TreeNode(stoi(val));
            q.push(parent->right);
        }
    }

    return root;
}

int main()
{
    string input;
    cout << "Enter the values of the nodes in level order, separated by spaces, use 'null' for no node: ";
    getline(cin, input);

    TreeNode *root = buildTree(input);
    vector<int> result = levelOrder(root);

    cout << "Level order traversal output: ";
    for (int val : result)
    {
        cout << val << " ";
    }

    return 0;
}

File: DSA/Pragya_Agarwal/Day15_Question2.py
class Node:

    def __init__(self, key):
        self.data = key
        self.left = None
        self.right = None

def levelOrder(root):
    h = height(root)
    for i in range(1, h+1):
        current_level(root, i)

def current_level(root, level):
    if root is None:
        return
    if level == 1:
        print(root.data, end=" ")
    elif level > 1:
        current_level(root.left, level-1)
        current_level(root.right, level-1)

def height(node):
    if node is None:
        return 0
    else:

        left_height = height(node.left)
        right_height = height(node.right)

        if left_height > right_height:
            return left_height+1
        else:
            return right_height+1


if __name__ == '__main__':
    root = Node(3)
    root.left = Node(9)
    root.right = Node(20)
    root.right.left = Node(15)
    root.right.right = Node(7)

    print("Level order traversal of binary tree is:")
    levelOrder(root)
